Feature/add branch selection

cnealen opened the pull request:
Summary of Workflow Logic Changes:

-Interactive Mode: Improved prompts, explicit selection sequences, and error handling.

-Non-Interactive Mode: Allows for fine-grained control with per-repo and global flags for build type, branch, source, and tag.

-Progress Tracking: Visually tracks long-running steps for user feedback, replacing previous unsupported characters for PW.

-Flexibility in Branches/Tags: Customization for feature builds and more intuitive default handling for development and release scenarios.

1. Progress Indicators and Spinner Animation
-Added a show_progress() function to display an animated spinner while long-running commands execute (git fetch/pull, docker save, singularity build).
-Replaced unsupported braille characters in progress tracker with a spinning bar.
-The spinner updates elapsed time every 0.5 seconds and shows a checkmark (✓) upon completion.
-All docker pull and docker build commands now run with this spinner (run_with_progress() wrapper).
-After completion, total time is displayed for each operation.

2. Branch and Tag Selection Improvements
-Feature Builds: When pulling images in feature build mode, the script now interactively prompts for which Docker tag to pull (not just :feature).
-Non-interactive mode supports --tag=REPO:TAG to specify which tag to pull per repo.
-Pulled images are automatically retagged as :feature for workflow consistency.
-Development builds now use the development branch for all repos automatically (no prompt).
-Release builds prompt for Docker tags but not for branches.

3. Per-Repository Branch and Source Selection
-You can set branch per repo using --branch=REPO:BRANCH in the command line.
--branch-default=BRANCH sets a global default branch.
-The new helper function get_repo_branch() selects the branch for each repo based on order of precedence (repo-specific > global default > build-type default).

4. User Experience Improvements
-The repo prompt now lists repos and asks users to select from those shown.
-The interactive workflow sequence is clarified: select build type, select repos, then source (build/pull), then branch/tag.
-Branch prompts for feature mode now occur only after image source selection.
-For ngen-forcing, a shared branch prompt for bmi/lumped/coastal variants.
- Verbose configuration summary removed for a cleaner interface.

5. Help and Usage
-Added a detailed --help flag explaining options, usage, and examples (including repo list, options for branches/tags, and command syntax).

6. Other Fixes
-Function definitions (like get_repo_branch) were reordered to avoid runtime errors due to undefined functions.

cnealen committed:
created branch selection option

cnealen committed:
added branch option and optimized code

cnealen committed:
added --help option

cnealen committed:
Major Changes:
Per-Repository Branch Selection: You can now specify different branches for each repository using:
--branch=REPO:BRANCH for specific repos (can be repeated)
--branch-default=BRANCH for a global default branch
Falls back to build-type default (development/release-candidate) if not specified
New Helper Function: Added get_repo_branch() function at line 490 that determines the branch to use based on priority:
Repo-specific branch (--branch=REPO:BRANCH)
Global default (--branch-default)
Build-type default (development/release-candidate)
Enhanced Interactive Mode: When running interactively, users are now prompted for:
Global default branch
Per-repo branch overrides
Image sources (build/pull)
Configuration Summary: Added a detailed summary display at line 354 showing:
Build type
Selected repos
Branch for each repo
Image source for each repo
Updated Help: The --help flag now shows the new branch options with examples
Example Usage:
# Different branches for different repos
./build_cluster.sh --build-type=development \
  --branch=ngen:feature/new-hydro \
  --branch=nwm-cal-mgr:bugfix/123 \
  ngen nwm-cal-mgr

# Global default with one override
./build_cluster.sh --build-type=development \
  --branch-default=feature/test \
  --branch=ngen:main \
  all

cnealen committed:
Problem: The script called get_repo_branch at line 380 (now 380) in the "Build Configuration" summary section, but the function wasn't defined until line 514, causing the error: ./build_cluster.sh: line 362: get_repo_branch: command not found Solution: I moved the get_repo_branch function definition from its original location (after line 509) to before the configuration summary section (now at lines 353-369), and removed the duplicate definition. The script should now work correctly when you run it on the ParallelWorks terminal. The function is now properly defined before its first use.

cnealen committed:
What was added:
show_progress() function (lines 442-463)
Displays an animated spinner (⠋⠙⠹⠸⠼⠴⠦⠧⠇⠏)
Shows elapsed time in MM:SS format
Updates every 0.5 seconds
Shows a checkmark (✓) when complete
run_with_progress() wrapper function (lines 465-489)
Runs any command with the progress indicator
Captures output and displays it after completion
Preserves exit status for error handling
Where it's now used:
Git operations: git fetch, git pull, git submodule update
Docker save: Saving Docker images to tar files
Singularity build: Converting tar archives to SIF files
Example output you'll see:
Updating ngen to latest from development branch...
[⠹] Fetching from origin (elapsed: 00:05)
And when complete:
[✓] Fetching from origin (completed in 00:07)

cnealen committed:
replaced 'release-candidate' with 'feature'

cnealen committed:
Updated the script so that when users select "development" build type in interactive mode, it will automatically use the 'development' branch for all repos without prompting for branch customization. The changes:
Line 282: Changed condition from != "release" to == "feature" - now only the "feature" build type will prompt for a global default branch
Line 322: Changed condition from != "release" to == "feature" - now only the "feature" build type will prompt for per-repo branch selection
Now the behavior is:
development: Automatically uses 'development' branch for all repos (no prompts)
release: Prompts for tags (no branch prompts)
feature: Prompts for branch customization (maintains flexibility)

cnealen committed:
Changes Made:
Repo selection prompt (line 286): Updated from "Enter repos to build (space-separated, or 'all'):" to "Enter repos to build (space-separated from the list above):"
Workflow order: The interactive flow now follows this exact sequence:
Select build type [1-3]
Enter repos to build
Display "Build type selected: {type}"
Display "Selected repos: {repos}"
Prompt for image source with format: "Image source for '{repo}' [build/pull] (default: {default}):"
For feature builds only: Prompt for branches AFTER image sources
Branch prompting (lines 333-347):
Now happens AFTER image source selection (not before)
Uses format: "Enter {repo} branch:" for most repos
Special handling for ngen-forcing: "Enter ngen-forcing branch (shared for bmi/lumped/coastal):"
Only prompts for feature builds in interactive mode
Removed verbose configuration summary: Removed the detailed "Build Configuration" section that displayed all branches and sources in a formatted box

cnealen committed:
Changes Made:
Added tag prompting for feature builds (build_cluster.sh:415-447)
When you select pull for a repo in feature build mode, the script now asks which Docker tag to pull
Similar to how release builds work, but only for repos set to pull mode
Updated pull logic to use specified tags (build_cluster.sh:930-935, build_cluster.sh:1023-1038)
Pulls the tag you specified instead of hardcoding :feature
Automatically retags the pulled image as :feature for consistency with the rest of the workflow
Added command-line argument support (build_cluster.sh:248-259)
You can now use --tag=REPO:TAG to specify tags non-interactively
Example: --tag=ngen-forcing:latest or --tag=ngen-forcing:development
How to Use:
Interactive mode (recommended):
./build_cluster.sh
# Select: 3) feature
# Enter: ngen-forcing
# Image source: pull
# When prompted: "Enter Docker tag to pull for ngen-forcing (shared for bmi/lumped): latest"
Non-interactive mode:
./build_cluster.sh --build-type=feature \
  --source=ngen-forcing:pull \
  --tag=ngen-forcing:latest \
  ngen-forcing
The script will now pull ghcr.io/ngwpc/ngen-bmi-forcing:latest (which exists) instead of trying to pull the non-existent :feature tag

cnealen committed:
changed progress tracker from unsupported braille characters to spinning bar

cnealen committed:
edited to remove conflicts

miguelp1986 reviewed (changes requested):
```
./build_cluster.sh --branch-default
````
No ouput. It should show error message

```
./build_cluster.sh --build-type=development --branch-default=my-feature \
      ngen nwm-cal-mgr ngen-forcing
```

Build-type development should not allow branch or branch default arguments. It should always be development branch for all repos. 

```
[ngen-pw-user@testrtxnwmdev2aws-1 nwm-automation-scripts]$ ./parallel_works_scripts/build_cluster.sh --build-type=development --branch=ngencerf-server:test
Available repos: ngencerf-ui ngencerf-server ngencerf-docker ngen nwm-cal-mgr ngen-forcing nwm-fcst-mgr nwm-verf
Enter repos to build (space-separated from the list above): ngencerf-server
Build type selected: development
Selected repos: ngencerf-server

Updating ngencerf-server to latest from test branch...
Fetching from origin
[✓] Fetching from origin (completed in 00:00)
error: pathspec 'test' did not match any file(s) known to git
Error: Branch 'test' does not exist in ngencerf-server
```

miguelp1986 reviewed (changes requested):
Looks like a bug
```
[ngen-pw-user@testrtxnwmdev2aws-1 nwm-automation-scripts]$ ./parallel_works_scripts/build_cluster.sh --build-type=feature --branch=ngen-forcing:sfincs_dockerfile_fix_mohammed ngen-forcing
Build type selected: feature
Selected repos: ngen-forcing
Image source for 'ngen-forcing' [build/pull] (default: pull): pull
Enter Docker tag to pull for ngen-forcing (shared for bmi/lumped): sfincs_dockerfile_fix_mohammed

./parallel_works_scripts/build_cluster.sh: line 1065: local: can only be used in a function
```

cnealen committed:
1. Fixed --branch-default with no value (build_cluster.sh:224-233)
Added validation to ensure --branch-default requires a value
Now shows: Error: --branch-default requires a value
Handles both --branch-default and --branch-default= cases
2. Prevent branch arguments with development build type (build_cluster.sh:286-298)
Added validation to reject --branch-default when --build-type=development
Added validation to reject --branch when --build-type=development
Shows clear error messages explaining that development builds always use the 'development' branch
3. Moved validation before directory creation (build_cluster.sh:280-298)
Validation now happens immediately after argument parsing
Prevents permission errors from occurring before argument validation

cnealen committed:
fixed the bug by removing local from three variable declarations that were outside of functions:
1. Line 983 - build_cluster.sh:983
Changed: local ngen_tag= → ngen_tag=
This variable is used when pulling ngen Docker images in feature builds
2. Lines 1086-1087 - build_cluster.sh:1086-1087
Changed: local pull_tag= → pull_tag=
Changed: local pull_image= → pull_image=
These variables are used when pulling Docker images for SIF conversion
The issue occurred because local can only be used inside bash functions.

miguelp1986 reviewed (changes requested):
No error message when Docker image tag is not found:

```
[ngen-pw-user@testrtxnwmdev2aws-1 parallel_works_scripts]$ ./build_cluster.sh --build-type=feature --branch=ngen-forcing:sfincs_dockerfile_fix_mohammed ngen-forcing
Build type selected: feature
Selected repos: ngen-forcing
Image source for 'ngen-forcing' [build/pull] (default: pull): pull
Enter Docker tag to pull for ngen-forcing (shared for bmi/lumped): sfincs_dockerfile_fix_mohammed ngen-forcing

Pulling docker image for SIF: ghcr.io/ngwpc/ngen-bmi-forcing:sfincs_dockerfile_fix_mohammed ngen-forcing
[✓] Pulling docker image for SIF: ghcr.io/ngwpc/ngen-bmi-forcing:sfincs_dockerfile_fix_mohammed ngen-forcing (completed in 00:00)
[ngen-pw-user@testrtxnwmdev2aws-1 parallel_works_scripts]$ echo $?
1
```

The exit code is 1, which means that an error occurred, but it was silenced. We should not silence errors. Make sure that all errors get routed properly to standard output/error.

miguelp1986 reviewed (changes requested):
We want to see all the output in the logs. The docker build logs are surpressed: 

```Selected repos: ngen-forcing
Image source for 'ngen-forcing' [build/pull] (default: pull):
Updating ngen-forcing to latest from sfincs_dockerfile_fix_mohammed branch...
Fetching from origin
^M[|] Fetching from origin (elapsed: 00:00)^M[/] Fetching from origin (elapsed: 00:01)^M[-] Fetching from origin (elapsed: 00:01)^M[✓] Fetching from origin (completed in 00:02)
From https://github.com/NGWPC/ngen-forcing
 * [new branch]      maxkipp-ngwpc-8941-upgrade-shapely-2 -> origin/maxkipp-ngwpc-8941-upgrade-shapely-2
 * [new branch]      philmiller-ebusy-failure -> origin/philmiller-ebusy-failure
Switched to a new branch 'sfincs_dockerfile_fix_mohammed'
branch 'sfincs_dockerfile_fix_mohammed' set up to track 'origin/sfincs_dockerfile_fix_mohammed'.
Pulling latest changes from sfincs_dockerfile_fix_mohammed
^M[|] Pulling latest changes from sfincs_dockerfile_fix_mohammed (elapsed: 00:00)^M[✓] Pulling latest changes from sfincs_dockerfile_fix_mohammed (completed in 00:00)
From https://github.com/NGWPC/ngen-forcing
 * branch            sfincs_dockerfile_fix_mohammed -> FETCH_HEAD
Already up to date.
Building ngen-bmi-forcing (feature) Docker image
^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:00)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:01)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:01)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:02)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:02)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:03)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:03)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:04)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:05)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:05)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:06)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:06)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:07)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:07)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:08)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:08)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:09)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:09)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:10)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:10)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:11)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:11)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:12)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:12)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:13)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:13)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:14)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:14)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:15)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:15)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:16)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:16)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:17)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:17)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:18)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:18)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:19)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:19)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:20)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:20)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:21)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:21)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:22)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:22)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:23)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:23)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:24)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:24)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:25)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:25)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:26)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:26)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:27)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:27)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:28)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:28)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:29)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:29)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:30)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:30)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:31)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:31)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:32)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:32)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:33)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:33)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:34)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:34)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:35)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:35)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:36)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:36)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:37)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:37)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:38)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:38)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:39)```

miguelp1986 commented on parallel_works_scripts/build_cluster.sh:
```diff
@@ -290,11 +559,11 @@ build_singularity_container_update_symlink() {
-    elif [[ "$build_type" == "release-candidate" ]]; then
-        sif_file="${sif_base}-release-candidate-${ts}.sif"
+    elif [[ "$build_type" == "feature" ]]; then
+        sif_file="${sif_base}-feature-${ts}.sif"
```
Use the actual feature branch name for the sif files.

miguelp1986 replied:
Make sure that the Docker image tags that are built locally also have the name of the branch as the tag and not just 'feature'.

miguelp1986 replied:
@cnealen This wasn't addressed with your updates as well.

miguelp1986 commented:
> We want to see all the output in the logs. The docker build logs are surpressed:
> 
> ```
> Image source for 'ngen-forcing' [build/pull] (default: pull):
> Updating ngen-forcing to latest from sfincs_dockerfile_fix_mohammed branch...
> Fetching from origin
> ^M[|] Fetching from origin (elapsed: 00:00)^M[/] Fetching from origin (elapsed: 00:01)^M[-] Fetching from origin (elapsed: 00:01)^M[✓] Fetching from origin (completed in 00:02)
> From https://github.com/NGWPC/ngen-forcing
>  * [new branch]      maxkipp-ngwpc-8941-upgrade-shapely-2 -> origin/maxkipp-ngwpc-8941-upgrade-shapely-2
>  * [new branch]      philmiller-ebusy-failure -> origin/philmiller-ebusy-failure
> Switched to a new branch 'sfincs_dockerfile_fix_mohammed'
> branch 'sfincs_dockerfile_fix_mohammed' set up to track 'origin/sfincs_dockerfile_fix_mohammed'.
> Pulling latest changes from sfincs_dockerfile_fix_mohammed
> ^M[|] Pulling latest changes from sfincs_dockerfile_fix_mohammed (elapsed: 00:00)^M[✓] Pulling latest changes from sfincs_dockerfile_fix_mohammed (completed in 00:00)
> From https://github.com/NGWPC/ngen-forcing
>  * branch            sfincs_dockerfile_fix_mohammed -> FETCH_HEAD
> Already up to date.
> Building ngen-bmi-forcing (feature) Docker image
> ^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:00)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:01)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:01)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:02)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:02)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:03)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:03)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:04)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:05)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:05)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:06)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:06)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:07)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:07)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:08)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:08)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:09)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:09)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:10)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:10)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:11)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:11)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:12)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:12)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:13)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:13)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:14)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:14)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:15)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:15)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:16)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:16)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:17)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:17)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:18)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:18)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:19)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:19)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:20)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:20)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:21)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:21)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:22)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:22)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:23)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:23)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:24)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:24)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:25)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:25)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:26)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:26)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:27)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:27)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:28)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:28)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:29)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:29)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:30)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:30)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:31)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:31)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:32)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:32)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:33)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:33)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:34)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:34)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:35)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:35)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:36)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:36)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:37)^M[/] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:37)^M[-] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:38)^M[\] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:38)^M[|] Building ngen-bmi-forcing (feature) Docker image (elapsed: 00:39)```
> ```

I actually see that the output gets logged after the Docker command completes, so we can skip this for now.

cnealen committed:
Improved run_with_progress() function (build_cluster.sh:522-554)
Now properly tracks the start time
Shows progress spinner while the command runs
Waits for the command to complete and captures the exit status
Displays the correct status symbol based on exit code
Shows error output from the command if available
Shows a fallback error message if the command fails but produces no output

miguelp1986 commented:
> No error message when Docker image tag is not found:
> 
> ```
> [ngen-pw-user@testrtxnwmdev2aws-1 parallel_works_scripts]$ ./build_cluster.sh --build-type=feature --branch=ngen-forcing:sfincs_dockerfile_fix_mohammed ngen-forcing
> Build type selected: feature
> Selected repos: ngen-forcing
> Image source for 'ngen-forcing' [build/pull] (default: pull): pull
> Enter Docker tag to pull for ngen-forcing (shared for bmi/lumped): sfincs_dockerfile_fix_mohammed ngen-forcing
> 
> Pulling docker image for SIF: ghcr.io/ngwpc/ngen-bmi-forcing:sfincs_dockerfile_fix_mohammed ngen-forcing
> [✓] Pulling docker image for SIF: ghcr.io/ngwpc/ngen-bmi-forcing:sfincs_dockerfile_fix_mohammed ngen-forcing (completed in 00:00)
> [ngen-pw-user@testrtxnwmdev2aws-1 parallel_works_scripts]$ echo $?
> 1
> ```
> 
> The exit code is 1, which means that an error occurred, but it was silenced. We should not silence errors. Make sure that all errors get routed properly to standard output/error.

@cnealen this hasn't been resolved with your changes. There is no tag `sfincs_dockerfile_fix_mohammed` for ngen-bmi-forcing so it should cause an error:

```
./build_cluster.sh --build-type=feature --branch=ngen-forcing:sfincs_dockerfile_fix_mohammed ngen-forcing
Build type selected: feature
Selected repos: ngen-forcing
Image source for 'ngen-forcing' [build/pull] (default: pull): pull
Enter Docker tag to pull for ngen-forcing (shared for bmi/lumped): sfincs_dockerfile_fix_mohammed

Pulling docker image for SIF: ghcr.io/ngwpc/ngen-bmi-forcing:sfincs_dockerfile_fix_mohammed
```

Here is the log file:

```
Build type selected: feature
Selected repos: ngen-forcing
Image source for 'ngen-forcing' [build/pull] (default: pull): Enter Docker tag to pull for ngen-forcing (shared for bmi/lumped): 
Pulling docker image for SIF: ghcr.io/ngwpc/ngen-bmi-forcing:sfincs_dockerfile_fix_mohammed
^M[|] Pulling docker image for SIF: ghcr.io/ngwpc/ngen-bmi-forcing:sfincs_dockerfile_fix_mohammed (elapsed: 00:00)
```

miguelp1986 commented:
These issues can be addressed later. We can now select any branch for each repo and successfully build it. Good job! I'll merge this.
